File: test/unit/pdr.test.js
import { describe, expect, it } from "vitest";
import { boxes, gpsFromChannels, parsePdrFile, series } from "../../public/pdr.js";
import { LAP_S, buildPdrDeltaMp4 } from "../fixtures/build.mjs";

describe("series", () => {
  const s = series([
    { t: 0, v: 0 },
    { t: 10, v: 100 },
    { t: 20, v: 300 },
  ]);

  it("interpolates values at a time", () => {
    expect(s.at(5)).toBe(50);
    expect(s.at(15)).toBe(200);
    expect(s.at(10)).toBe(100);
  });

  it("inverts monotonic series with timeAt", () => {
    expect(s.timeAt(50)).toBe(5);
    expect(s.timeAt(200)).toBe(15);
  });

  it("computes a central-difference rate", () => {
    // between t=10 and t=20 the slope is 20 v/t
    expect(s.rate(15, 2)).toBe(20);
  });

  it("exposes first/last and length", () => {
    expect(s.n).toBe(3);
    expect(s.first).toEqual({ t: 0, v: 0 });
    expect(s.last).toEqual({ t: 20, v: 300 });
  });
});

describe("gpsFromChannels", () => {
  const t = (i) => i * 0.15;
  const chan = (n, deg) => Array.from({ length: n }, (_, i) => ({ t: t(i), v: Math.round(deg(i) * 1e7) }));

  it("decodes deg*1e7 lat/lon channels into a degrees trace", () => {
    const lat = chan(40, (i) => 36.56 + i * 1e-4);
    const lon = chan(40, (i) => -79.2 + i * 1e-4);
    const gps = gpsFromChannels(lat, lon);
    expect(gps).toHaveLength(40);
    expect(gps[10].lat).toBeCloseTo(36.561, 6);
    expect(gps[10].lon).toBeCloseTo(-79.199, 6);
    expect(gps[10].v).toBeUndefined();
  });

  it("takes speed from the odometer series when given", () => {
    const lat = chan(40, (i) => 36.56 + i * 1e-4);
    const lon = chan(40, (i) => -79.2 + i * 1e-4);
    const odo = series(Array.from({ length: 40 }, (_, i) => ({ t: t(i), v: i * 6 }))); // 40 m/s
    const gps = gpsFromChannels(lat, lon, odo);
    expect(gps[20].v).toBeCloseTo(40, 6);
  });

  it("decodes float32-bit lat/lon channels via the fallback interpretation", () => {
    const f32 = new DataView(new ArrayBuffer(4));
    const asBits = (deg) => {
      f32.setFloat32(0, deg);
      return f32.getInt32(0);
    };
    const lat = Array.from({ length: 40 }, (_, i) => ({ t: t(i), v: asBits(36.56 + i * 1e-4) }));
    const lon = Array.from({ length: 40 }, (_, i) => ({ t: t(i), v: asBits(-79.2 + i * 1e-4) }));
    const gps = gpsFromChannels(lat, lon);
    expect(gps[10].lat).toBeCloseTo(36.561, 4);
    expect(gps[10].lon).toBeCloseTo(-79.199, 4);
  });

  it("returns null rather than a garbage trace", () => {
    // parked car: zero extent
    const stuck = Array.from({ length: 40 }, (_, i) => ({ t: t(i), v: 365600000 }));
    expect(gpsFromChannels(stuck, stuck)).toBeNull();
    // values plausible under neither interpretation
    const noise = Array.from({ length: 40 }, (_, i) => ({ t: t(i), v: 2000000000 - i * 55555555 }));
    expect(gpsFromChannels(noise, noise)).toBeNull();
    // too few samples
    expect(gpsFromChannels([], [])).toBeNull();
  });
});

describe("parsePdrFile with a delta-encoded stream (real firmware shape)", () => {
  const lapS = LAP_S(); // 47.12s
  const parse = (opts) => parsePdrFile(new Blob([buildPdrDeltaMp4(opts)]));

  it("decodes the delta-encoded GPS channels into a dictionary-scaled trace", async () => {
    const out = await parse();
    expect(out.gps).not.toBeNull();
    expect(out.gps.length).toBeGreaterThan(100);
    // ~2Hz stream around the reference circle at lat0/lon0
    const lats = out.gps.map((p) => p.lat);
    const lons = out.gps.map((p) => p.lon);
    expect(Math.min(...lats)).toBeGreaterThan(36.5545);
    expect(Math.max(...lats)).toBeLessThan(36.5655);
    expect((Math.min(...lons) + Math.max(...lons)) / 2).toBeCloseTo(-79.2, 2);
    // racing-line speed comes from the Speed channel (m/s), modulated ±5% around 40
    const vs = out.gps.map((p) => p.v);
    expect(Math.max(...vs)).toBeGreaterThan(40);
    expect(Math.max(...vs)).toBeLessThan(42.5);
  });

  it("keeps decoder state across sample boundaries", async () => {
    // the fixture splits records into 250-record samples; a state reset would
    // orphan every delta at a sample start and thin or corrupt the trace
    const out = await parse();
    const dt = out.gps.slice(1).map((p, i) => p.t - out.gps[i].t);
    expect(Math.max(...dt)).toBeLessThan(1.5); // no holes
    expect(out.durationS).toBeGreaterThan(150);
  });

  it("times laps from delta-stream beacons and reads mrlv date/time", async () => {
    const out = await parse({ beaconTimes: [30, 30 + lapS, 30 + 2 * lapS] });
    const ms = Math.round(lapS * 1000); // 47124
    expect(out.laps.map((l) => l.timeMs)).toEqual([ms, ms]);
    expect(out.laps.every((l) => !l.estimated)).toBe(true);
    expect(out.date).toBe("2026-06-20");
    expect(out.time).toBe("09:15:00");
  });

  it("reports top speed, max RPM and max lateral G from the car channels", async () => {
    const out = await parse();
    // speed peaks at 42 m/s = 151.2 km/h; rpm at 6000; latAcc at v²/r ≈ 0.6 G
    expect(out.metrics.topSpeedKph).toBeGreaterThan(148);
    expect(out.metrics.topSpeedKph).toBeLessThan(152);
    expect(out.metrics.maxRpm).toBeGreaterThan(5900);
    expect(out.metrics.maxRpm).toBeLessThanOrEqual(6000);
    expect(out.metrics.maxLatG).toBeGreaterThan(0.5);
    expect(out.metrics.maxLatG).toBeLessThan(0.65);
  });
});

describe("boxes", () => {
  // Build a buffer holding two MP4 boxes: "ftyp" (12 bytes) and "free" (8 bytes).
  function buildBoxes() {
    const buf = new ArrayBuffer(20);
    const dv = new DataView(buf);
    const put = (off, size, type) => {
      dv.setUint32(off, size);
      [...type].forEach((ch, i) => dv.setUint8(off + 4 + i, ch.charCodeAt(0)));
    };
    put(0, 12, "ftyp");
    put(12, 8, "free");
    return dv;
  }

  it("parses consecutive box headers", () => {
    const dv = buildBoxes();
    const out = boxes(dv, 0, 20);
    expect(out).toEqual([
      { type: "ftyp", start: 0, body: 8, size: 12 },
      { type: "free", start: 12, body: 20, size: 8 },
    ]);
  });

  it("stops at garbage instead of running away", () => {
    const dv = buildBoxes();
    dv.setUint32(12, 3); // invalid size < 8
    expect(boxes(dv, 0, 20)).toHaveLength(1);
  });
});

File: apps/android/core/src/test/kotlin/app/trackevolution/core/telemetry/PDRTest.kt
package app.trackevolution.core.telemetry

import app.trackevolution.core.JsMath
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.nio.ByteBuffer

/**
 * `test/unit/pdr.test.js`, ported case for case.
 *
 * The JS builds its delta-encoded input with `buildPdrDeltaMp4()`; the same
 * bytes are committed as `contracts/logic/video/pdr-delta.mp4` (with beacons)
 * and `pdr-delta-nobeacon.mp4` (without), so the Kotlin can ask the identical
 * question of the identical file.
 */
class PDRTest {

    // ---- gpsFromChannels ----------------------------------------------------------

    private fun t(i: Int) = i * 0.15
    private fun chan(n: Int, deg: (Int) -> Double) =
        List(n) { i -> ChannelPoint(t = t(i), v = JsMath.round(deg(i) * 1e7, 1.0)) }

    @Test
    fun `decodes deg times 1e7 lat and lon channels into a degrees trace`() {
        val lat = chan(40) { 36.56 + it * 1e-4 }
        val lon = chan(40) { -79.2 + it * 1e-4 }
        val gps = given(PDR.gpsFromChannels(lat, lon))
        assertEquals(40, gps.size)
        assertEquals(36.561, gps[10].lat, 1e-6)
        assertEquals(-79.199, gps[10].lon, 1e-6)
        assertNull(gps[10].v)
    }

    @Test
    fun `takes speed from the odometer series when given`() {
        val lat = chan(40) { 36.56 + it * 1e-4 }
        val lon = chan(40) { -79.2 + it * 1e-4 }
        val odo = series(List(40) { i -> ChannelPoint(t = t(i), v = i * 6.0) }) // 40 m/s
        val gps = given(PDR.gpsFromChannels(lat, lon, odo))
        assertEquals(40.0, given(gps[20].v), 1e-6)
    }

    @Test
    fun `decodes float32-bit lat and lon channels via the fallback interpretation`() {
        fun asBits(deg: Double): Double = java.lang.Float.floatToIntBits(deg.toFloat()).toDouble()
        val lat = List(40) { i -> ChannelPoint(t = t(i), v = asBits(36.56 + i * 1e-4)) }
        val lon = List(40) { i -> ChannelPoint(t = t(i), v = asBits(-79.2 + i * 1e-4)) }
        val gps = given(PDR.gpsFromChannels(lat, lon))
        assertEquals(36.561, gps[10].lat, 1e-4)
        assertEquals(-79.199, gps[10].lon, 1e-4)
    }

    @Test
    fun `returns null rather than a garbage trace`() {
        // parked car: zero extent
        val stuck = List(40) { i -> ChannelPoint(t = t(i), v = 365600000.0) }
        assertNull(PDR.gpsFromChannels(stuck, stuck))
        // values plausible under neither interpretation
        val noise = List(40) { i -> ChannelPoint(t = t(i), v = 2000000000.0 - i * 55555555.0) }
        assertNull(PDR.gpsFromChannels(noise, noise))
        // too few samples
        assertNull(PDR.gpsFromChannels(emptyList(), emptyList()))
    }

    // ---- parsePdrFile with a delta-encoded stream (real firmware shape) ---------

    private val lapS = 2 * Math.PI * 300 / 40 // 47.12s

    @Test
    fun `decodes the delta-encoded GPS channels into a dictionary-scaled trace`() {
        val out = VideoFixtures.parse("pdr-delta-nobeacon.mp4")
        val gps = given(out.gps)
        assertTrue(gps.size > 100)
        // ~2Hz stream around the reference circle at lat0/lon0
        assertTrue(gps.minOf { it.lat } > 36.5545)
        assertTrue(gps.maxOf { it.lat } < 36.5655)
        assertEquals(-79.2, (gps.minOf { it.lon } + gps.maxOf { it.lon }) / 2, 0.005)
        // racing-line speed comes from the Speed channel (m/s), modulated ±5% around 40
        val vs = gps.map { it.v ?: 0.0 }
        assertTrue(vs.max() > 40)
        assertTrue(vs.max() < 42.5)
    }

    @Test
    fun `keeps decoder state across sample boundaries`() {
        // the fixture splits records into 250-record samples; a state reset would
        // orphan every delta at a sample start and thin or corrupt the trace
        val out = VideoFixtures.parse("pdr-delta-nobeacon.mp4")
        val gps = given(out.gps)
        val dt = gps.drop(1).mapIndexed { i, p -> p.t - gps[i].t }
        assertTrue(dt.max() < 1.5) // no holes
        assertTrue(out.durationS > 150)
    }

    @Test
    fun `times laps from delta-stream beacons and reads mrlv date and time`() {
        val out = VideoFixtures.parse("pdr-delta.mp4")
        val ms = JsMath.roundToInt(lapS * 1000) // 47124
        assertEquals(listOf(ms, ms), out.laps.map { it.timeMs })
        assertTrue(out.laps.all { !it.estimated })
        assertEquals("2026-06-20", out.date)
        assertEquals("09:15:00", out.time)
    }

    @Test
    fun `reports top speed, max RPM and max lateral G from the car channels`() {
        val out = VideoFixtures.parse("pdr-delta-nobeacon.mp4")
        val metrics = given(out.metrics)
        // speed peaks at 42 m/s = 151.2 km/h; rpm at 6000; latAcc at v²/r ≈ 0.6 G
        val top = given(metrics.topSpeedKph)
        assertTrue(top > 148 && top < 152)
        val rpm = given(metrics.maxRpm)
        assertTrue(rpm > 5900 && rpm <= 6000)
        val g = given(metrics.maxLatG)
        assertTrue(g > 0.5 && g < 0.65)
    }

    @Test
    fun `decodes throttle, brake and steering car channels`() {
        val out = VideoFixtures.parse("pdr-delta-nobeacon.mp4")
        // pedals are raw 0-255 scaled by the dictionary's "%" units to 0-100,
        // alternating (throttle on the sine's positive half, brake the negative)
        val th = given(out.carChannels.throttle).map { it.v }
        val br = given(out.carChannels.brake).map { it.v }
        assertEquals(0.0, th.min())
        assertTrue(th.max() > 99)
        assertTrue(th.max() <= 100)
        assertTrue(br.max() > 99)
        // steering is stored in radians with an empty units string (the real
        // firmware shape); the parser converts to degrees itself: ±0.5 rad = ±28.6°
        val st = given(out.carChannels.steering).map { it.v }
        assertTrue(st.max() > 28)
        assertTrue(st.max() < 29.2)
        assertTrue(st.min() < -28)
    }

    // ---- boxes -----------------------------------------------------------------------

    /** Two MP4 boxes: "ftyp" (12 bytes) and "free" (8 bytes). */
    private fun buildBoxes(): ByteArray {
        val buf = ByteBuffer.allocate(20)
        fun put(off: Int, size: Int, type: String) {
            buf.putInt(off, size)
            type.forEachIndexed { i, ch -> buf.put(off + 4 + i, ch.code.toByte()) }
        }
        put(0, 12, "ftyp")
        put(12, 8, "free")
        return buf.array()
    }

    @Test
    fun `parses consecutive box headers`() {
        val out = MP4.boxes(ByteView(buildBoxes()), 0, 20)
        assertEquals(
            listOf(
                MP4.Box(type = "ftyp", start = 0, body = 8, size = 12),
                MP4.Box(type = "free", start = 12, body = 20, size = 8),
            ),
            out,
        )
    }

    @Test
    fun `stops at garbage instead of running away`() {
        val bytes = buildBoxes()
        ByteBuffer.wrap(bytes).putInt(12, 3) // invalid size < 8
        assertEquals(1, MP4.boxes(ByteView(bytes), 0, 20).size)
    }
}
